Filter and Tag Filter plugins - Not live, "Or" conditions, Allow Filter UI in column

fsnoddy opened the pull request:
…(Can be useful for AA tracking)

- Add an "Or" option for different groupings
- Modify Filter max width to be in pixels so that it can be nicely nested in a column

_(Pick the language of your choice to fill out the following information. / Choisissez la langue de votre choix pour remplir l'information ci-dessous.)_

## What does this pull request (PR) do? / Que fait cette demande « pull » (PR)?
A clear and concise description of what the code of this PR accomplishes. /// Une description claire et concise de ce que le code de cette PR accompli.

## Additional information (optional) / Information additionnelle (optionel)

**General checklist / Liste de contrôle générale**
Make your own list for the purpose of your Pull request. /// Faites votre propre liste en fonction des besoins de votre demande « pull ».

- [x] Create/update documentation /// Créer/mettre à jour la documentation
- [x] Build and test PR's code /// Rouler le script de compilation et tester le code de la PR
- [ ] Validate changes against WCAG for accessibility /// Valider les changements avec WCAG pour l'accessibilité
- [ ] Ensure documentation is bilingual /// S'assurer que la documentation soit bilingue

**Related issues / Requêtes associées**
List issues that are being closed or worked on with this pull request i.e. Closes #8941 /// Listez les autres requêtes (« issues » ou PR) qui sont fermées ou traitées avec cette demande de retrait ex. : Closes #8941

**Screenshots / Captures d'écrans**
If applicable, add screenshots to help demonstrate what this PR does. /// Si applicable, ajoutez des captures d'écran pour aider à démontrer ce que cette PR fait.

fsnoddy committed:
- Add option to not filter live on the Filter and Tag Filter plugins (Can be useful for AA tracking)
- Add an "Or" option for different groupings
- Modify Filter max width to be in pixels so that it can be nicely nested in a column

fsnoddy committed:
- Cleanup PR
- Fix french example
- Remove testing code

fsnoddy committed:
Fix eslint issues.

fsnoddy committed:
Fix eslint issues.

fsnoddy committed:
Merge branch 'master' of https://github.com/fsnoddy/wet-boew

fsnoddy committed:
Aria not supported on <form> element

duboisp reviewed (changes requested):
Partially reviewed, still need a full review + local testing.

This is a **Minor** change to "Filter" plugin and a **Patch** to "Tag Filter" provisional plugin

The following CI error has been raised with HTMLlint :

```
Running "htmllint:all" (htmllint) task
dist/unmin/demos/tagfilter/tagfilter-en.html [L1107:C92] Element “search” not allowed as child of element “div” in this context. (Suppressing further errors from this subtree.)
dist/unmin/demos/tagfilter/tagfilter-en.html [L1107:C92] Possible misuse of “aria-label”. (If you disagree with this warning, file an issue report or send e-mail to www-validator@w3.org.)
dist/unmin/demos/tagfilter/tagfilter-fr.html [L1127:C92] Element “search” not allowed as child of element “div” in this context. (Suppressing further errors from this subtree.)
dist/unmin/demos/tagfilter/tagfilter-fr.html [L1127:C92] Possible misuse of “aria-label”. (If you disagree with this warning, file an issue report or send e-mail to www-validator@w3.org.)
```

fsnoddy commented on src/plugins/filter/_screen-sm-min.scss:
```diff
@@ -1,6 +1,6 @@
 .wb-filter {
 	.input-group {
-		max-width: 80%;
+		max-width: 600px;
```
Percentage is problematic in the context of a limited width container (example in a col-md-3, it will take 80% of that column). Like in the "Finder" example added in TagFilter

You might have a more elegant solution but like the rule targeting something in a col-* for example.

fsnoddy committed:
Very minor change, fixed "that" to "self" as a best practice. I believe I had seen it in another plugin which is why this nomenclature had been used.

fsnoddy commented on src/plugins/filter/filter.js:
```diff
@@ -217,6 +219,7 @@ var componentName = "wb-filter",
 			hndParentSelector = settings.hdnparentuntil,
 			$items = $elm.find( secSelector + settings.selector ),
 			itemsLength = $items.length,
+			that = this || $field.get( 0 ),
```
This change was made. 

CI error is based on erroneous NU Validator. <search> element should be valid inside any flow content like div, section, main, etc.. All those currently show up as errors as described in [this issue](https://github.com/validator/validator/issues/1569). 

[Browsers seem to show pretty decent support for it](https://caniuse.com/mdn-html_elements_search).

duboisp commented on src/plugins/tagfilter/tagfilter-en.hbs:
```diff
@@ -981,6 +982,246 @@
+	<h2>Example 4 - Finder with "or" behavior updating with a button</h2>
+
+	<div class="row">
+	<search id="searchFilter2" class="col-xs-12 col-md-3 pull-right" aria-label="On this page">
```
Why there is an `aria-label` here? There is already a few heading preceding and inside the search section.

fsnoddy replied:
Very fair point, there is likely already enough context here without the aria-label. Change was made with new commit.

Was just making sure it wasn't conflicting with the site-wide search element which also has a aria-role="search".

duboisp reviewed (changes requested):
Can you add an exception for the error about the `<search>` element? in the ignore list here: https://github.com/wet-boew/wet-boew/blob/master/Gruntfile.coffee#L952-L961

fsnoddy committed:
Exception added for search element to address linting issue until the Nu validator's issue is fixed.

fsnoddy committed:
Aria label not necessary since the heading already adds context to the element.

Garneauma reviewed (changes requested):
@fsnoddy Please rebase this pull request and fix the conflicts. Also, please run `grunt dist` locally. There are a few HTML issues that need to be fixed.

fsnoddy committed:
Merge branch 'master' into tagFilterUpdates

fsnoddy committed:
Lint issues

fsnoddy committed:
Update tagfilter.js

fsnoddy commented:
Is there something I can do to help this forward? I know you guys are extremely busy.

Garneauma commented:
@fsnoddy The build is still failing. Can you make sure the build passes? If you are not able to do so, we'll fix it for you when we have the bandwidth. Thank you!

EricDunsworth commented on Gruntfile.coffee:
```diff
@@ -957,6 +957,9 @@ module.exports = (grunt) ->
 						"The “contentinfo” role is unnecessary for element “footer”."
 						"The “navigation” role is unnecessary for element “nav”."
 						"The “banner” role is unnecessary for element “header”."
+						"Element “search” not allowed as child of element “main” in this context."
+						"Element “search” not allowed as child of element “section” in this context."
+						"Element “search” not allowed as child of element “div” in this context."
```
Here's a fix for the ``htmllint:all`` task's failures :).

Needed an extra part at the end of the error messages.

```suggestion
						"Element “search” not allowed as child of element “main” in this context. (Suppressing further errors from this subtree.)"
						"Element “search” not allowed as child of element “section” in this context. (Suppressing further errors from this subtree.)"
						"Element “search” not allowed as child of element “div” in this context. (Suppressing further errors from this subtree.)"
```

EricDunsworth commented on src/plugins/tagfilter/tagfilter-en.hbs:
```diff
@@ -981,6 +982,246 @@
+				<legend>Audience</legend>
+
+				<ul class="list-unstyled">
+					<li class="checkbox"><label><input type="checkbox" name="audience" value="student" class="wb-tagfilter-ctrl" data-wb-tagfilter='{"group": "orConditionGroup"}' /> Student</label></li>
+					<li class="checkbox"><label><input type="checkbox" name="audience" value="worker" class="wb-tagfilter-ctrl" data-wb-tagfilter='{"group": "orConditionGroup"}' /> Worker</label></li>
+					<li class="checkbox"><label><input type="checkbox" name="audience" value="senior" class="wb-tagfilter-ctrl" data-wb-tagfilter='{"group": "orConditionGroup"}' /> Senior</label></li>
```
The ``bootlint`` task is failing on this checkbox structure (see [error E018](https://github.com/twbs/bootlint/wiki/E018)).

**Possible workarounds:**
* Use  a group of ``div``s instead of a list
* Remove ``class="checkbox"`` from the ``li`` elements and wrap ``<div class="checkbox">...</div>`` around the labels.
* Modify the [``bootlint:all`` task](https://github.com/wet-boew/wet-boew/blob/01060c7bb1e641afbdc7f1f313c276eef294eb68/Gruntfile.coffee#L974-L994) to ignore the error on the tag filter plugin's demo pages
  ```coffee
  "E018" # Incorrect markup used with the `.radio` class. The correct markup structure is `.radio>label>input[type="radio"]`
  ```
  **Note:** Might want to create a separate ``tagfilter`` bootlint task to avoid exempting it globally... btw I was unable to get [``relaxerror``'s file targeting feature](https://www.npmjs.com/package/grunt-bootlint#optionsrelaxerror) working correctly in the ``all`` task :S

**PS:**
[Bootlint is archived](https://github.com/twbs/bootlint)... but still technically makes sense to have in WET/GCWeb given that both are based on ancient versions of Bootstrap :P.

EricDunsworth replied:
Whoops - nevermind. I misinterpreted what was up. Please ignore this.

EricDunsworth commented on src/plugins/tagfilter/tagfilter-fr.hbs:
```diff
@@ -1000,6 +1001,254 @@
+				<legend>Audience</legend>
+
+				<ul class="list-unstyled">
+					<li class="radio"><label><input type="checkbox" name="audience" value="student" class="wb-tagfilter-ctrl" data-wb-tagfilter='{"group": "orConditionGroup"}' /> Étudiant</label></li>
+					<li class="radio"><label><input type="checkbox" name="audience" value="worker" class="wb-tagfilter-ctrl" data-wb-tagfilter='{"group": "orConditionGroup"}' /> Travailleur</label></li>
+					<li class="radio"><label><input type="checkbox" name="audience" value="senior" class="wb-tagfilter-ctrl" data-wb-tagfilter='{"group": "orConditionGroup"}' /> Aîné</label></li>
```
The bootlint task is freaking out because of the "radio" class being used with checkbox inputs (see [error E018](https://github.com/twbs/bootlint/wiki/E018)). So let's switch over to the "checkbox" class.

Btw EN isn't affected (it's already using the "checkbox" class).

```suggestion
					<li class="checkbox"><label><input type="checkbox" name="audience" value="student" class="wb-tagfilter-ctrl" data-wb-tagfilter='{"group": "orConditionGroup"}' /> Étudiant</label></li>
					<li class="checkbox"><label><input type="checkbox" name="audience" value="worker" class="wb-tagfilter-ctrl" data-wb-tagfilter='{"group": "orConditionGroup"}' /> Travailleur</label></li>
					<li class="checkbox"><label><input type="checkbox" name="audience" value="senior" class="wb-tagfilter-ctrl" data-wb-tagfilter='{"group": "orConditionGroup"}' /> Aîné</label></li>
```

EricDunsworth commented on src/plugins/tagfilter/tagfilter-fr.hbs:
```diff
@@ -1000,6 +1001,254 @@
+				<legend>Thème</legend>
+
+				<ul class="list-unstyled">
+					<li class="radio"><label><input type="checkbox" name="stage" value="divorced" class="wb-tagfilter-ctrl" data-wb-tagfilter='{"group": "orConditionGroup"}' /> Divorcé ou séparé</label></li>
+					<li class="radio"><label><input type="checkbox" name="stage" value="retirement" class="wb-tagfilter-ctrl" data-wb-tagfilter='{"group": "orConditionGroup"}' /> Retraite</label></li>
+					<li class="radio"><label><input type="checkbox" name="stage" value="family" class="wb-tagfilter-ctrl" data-wb-tagfilter='{"group": "orConditionGroup"}' /> Fonder une famille</label></li>
```
Ditto.

```suggestion
					<li class="checkbox"><label><input type="checkbox" name="stage" value="divorced" class="wb-tagfilter-ctrl" data-wb-tagfilter='{"group": "orConditionGroup"}' /> Divorcé ou séparé</label></li>
					<li class="checkbox"><label><input type="checkbox" name="stage" value="retirement" class="wb-tagfilter-ctrl" data-wb-tagfilter='{"group": "orConditionGroup"}' /> Retraite</label></li>
					<li class="checkbox"><label><input type="checkbox" name="stage" value="family" class="wb-tagfilter-ctrl" data-wb-tagfilter='{"group": "orConditionGroup"}' /> Fonder une famille</label></li>
```

fsnoddy committed:
Update Gruntfile.coffee

Co-authored-by: EricDunsworth <1907279+EricDunsworth@users.noreply.github.com>

fsnoddy committed:
Update src/plugins/tagfilter/tagfilter-fr.hbs

Co-authored-by: EricDunsworth <1907279+EricDunsworth@users.noreply.github.com>

fsnoddy committed:
Update src/plugins/tagfilter/tagfilter-fr.hbs

Co-authored-by: EricDunsworth <1907279+EricDunsworth@users.noreply.github.com>

Garneauma commented on src/plugins/filter/_screen-lg-min.scss:
```diff
@@ -1,6 +1,6 @@
 .wb-filter {
 	.input-group {
-		max-width: 60%;
+		max-width: 702px;
```
@duboisp The percentage does not work because when you wrap the filter in a smaller column, the search bar becomes very small when it should take the full-width of the column. The has been annoying for a while, but never got around to fix it. Thank you @fsnoddy for this.

Garneauma commented on site/pages/docs/ref/filter/filter-en.hbs:
```diff
@@ -203,6 +212,12 @@
+				<td><code>live</code></td>
+				<td>Define if the filtering should happen when the keyboard is typed or on submit of form element inside the plugin container.</td>
+				<td><code>data-wb-filter='{"live": "<em>false</em>"}'</code></td>
+				<td>True by default. False will update the results only when a submit happens on the form. Can be useful for Adobe Analytics tracking.</td>
```
```suggestion
				<td>
					<dl>
						<dt>True (default):</dt>
						<dd>Will update the results on key up.</dd>
						<dt>False</dt>
						<dd>Will update the results only when a submit happens on the form. Can be useful for Adobe Analytics tracking.</dd>
					</dl>
				</td>
```

Garneauma reviewed (changes requested):
Thank you for your contribution @fsnoddy, please see below the requested changes and comments:

For the Tagfilter plugin:
- Please add the "or" feature to the documentation. 
  - Note: is there any way to apply the "group" option to the fieldset (if there is one) instead of each individual field for checkboxes and radio buttons? Just a suggestion. It might be easier for authors to implement.
- To be discussed: calling the option "onSubmit" might be more descriptive.
- Please add a "Configuration options" section at the bottom of the documentation page with all the options. See the [Filter plugin documentation](https://wet-boew.github.io/wet-boew/docs/ref/filter/filter-en.html) for an example.

For the Filter plugin:
- Please add a working example for the "live" feature. Please also make sure it still works when using the "uiTemplate" option.
- To be discussed: calling the option "onSubmit" might be more descriptive.

fsnoddy committed:
Update site/pages/docs/ref/filter/filter-en.hbs

Co-authored-by: Garneauma <garneauma@gmail.com>

fsnoddy committed:
Merge branch 'wet-boew:master' into tagFilterUpdates

fsnoddy committed:
- Changed the french documentation to english as requested.
- Added group option to fieldset
- Added tag filter configurations options
- Added example for the Filter plugin

fsnoddy commented on site/pages/docs/ref/filter/filter-fr.hbs:
```diff
@@ -60,6 +60,15 @@
 	&lt;li&gt;Alberta&lt;/li&gt;
 	...
 &lt;/ul&gt;</code></pre>
+		</li>
```
All of the requested changes have been added to the pull request. 

Tag filter
1. Or documentation was added;
2. Configuration options section added;
3. Users can now set the option on Either a parent fieldset or on specific input;
4. "live" property was used to keep consistency with the option available for Fieldflow

Filter

1. Example was added;
2. Users get a console error if they use the live false with a custom UI and no button.
3. If the button is automatically added if no custom UI is provided.
4. "live" property was used to keep consistency with the option available for Fieldflow

Garneauma commented:
@fsnoddy Could you please rebase your PR to the current head of the upstream/master branch?
